pkg/controller/garbagecollector/rate_limiter_helper.go

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package garbagecollector
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/typed/dynamic"
+	"k8s.io/kubernetes/pkg/util/metrics"
+)
+
+// RegisteredRateLimiter records the registered RateLimters to avoid
+// duplication.
+type RegisteredRateLimiter struct {
+	rateLimiters map[unversioned.GroupVersion]struct{}
+	lock         sync.Mutex
+}
+
+// NewRegisteredRateLimiter returns a new RegisteredRateLimiater.
+func NewRegisteredRateLimiter() *RegisteredRateLimiter {
+	return &RegisteredRateLimiter{
+		rateLimiters: make(map[unversioned.GroupVersion]struct{}),
+		lock:         sync.Mutex{},
+	}
+}
+
+func (r *RegisteredRateLimiter) register(gv unversioned.GroupVersion, client *dynamic.Client, prefix string) {
```

Comment:
The write only occurs once, and it occurs at the beginning, so there should be a way to avoid using a RLock in the later reads.

If the variable we want to protect is an `int`, I think we can do it this way:
``` go
var x int // the variable to be protected

func (r *RegisteredRateLimiter) registerIfNotPresent() {
  if x == 0 {
    r.Lock()
    if x == 0 {
      x = 1
    }
    r.Unclok()
  }
}
```

But in reality instead of a `var x int`, we are using a `map[unversioned.GroupVersion]struct{}`, and cocurrent read/write to a map causes [fatal error](https://github.com/go-zoo/bone/issues/39#issue-138039113
) in Go 1.6, so the above code won't work...